Sources/audio-devices/AudioDevices.swift

```diff
@@ -106,6 +107,92 @@ struct AudioDevice: Hashable, Codable, Identifiable {
     )
   }
 
+  var isInputMuted: Bool? {
+    do {
+      var muteValue: UInt32 = 0
+      try CoreAudioData.get(
+        id: id,
+        selector: kAudioDevicePropertyMute,
+        scope: kAudioDevicePropertyScopeInput,
+        value: &muteValue
+      )
+      
+      return muteValue != 0
+    } catch {
+      return nil
+    }
+  }
+  
+  var isOutputMuted: Bool? {
+    do {
+      var muteValue: UInt32 = 0
+      try CoreAudioData.get(
+        id: id,
+        selector: kAudioDevicePropertyMute,
+        scope: kAudioDevicePropertyScopeOutput,
+        value: &muteValue
+      )
+      
+      return muteValue != 0
+    } catch {
+      return nil
+    }
+  }
+
+  var isMuted: Bool? {
+    guard !(isInput && isOutput) else {
+      return nil
+    }
+
+    if isInput {
+      return self.isInputMuted
+    }
+    if isOutput {
+      return self.isOutputMuted
+    }
+
+    return nil
+  }
+
+  func toggleInputMute() throws {
+    guard let isInputMuted = self.isInputMuted else {
+      throw AudioDevice.Error.muteNotSupported
+    }
+    
+    var newValue = NSNumber(booleanLiteral: !isInputMuted).uint32Value
+    
+    try CoreAudioData.set(
+      id: id,
+      selector: kAudioDevicePropertyMute,
+      scope: kAudioDevicePropertyScopeInput,
+      value: &newValue
+    )
+  }
+  
+  func toggleOutputMute() throws {
+    guard let isOutputMuted = self.isOutputMuted else {
+      throw AudioDevice.Error.muteNotSupported
+    }
+    
+    var newValue = NSNumber(booleanLiteral: !isOutputMuted).uint32Value
+    
+    try CoreAudioData.set(
+      id: id,
+      selector: kAudioDevicePropertyMute,
+      scope: kAudioDevicePropertyScopeOutput,
+      value: &newValue
+    )
+  }
+
+  func toggleMute() throws {
+    if isInput {
```

Comment:
In your previous PR you'd left a comment indicating you'd need to work out how to handle this for devices which have input and output. Has this been addressed by defaulting to input? How would you mute the audio of a device with both input and output?

Perhaps it'd be prudent to change the api to `toggleInputMute` and `toggleOutputMute`, and perhaps `toggleMute` could do both/whichever is present? Another option maybe to default similarly to current implementation but accept an argument which allows the user to specify input/output?

Reply:
Having input/output as a parameter will require to add enum only for this purpose. On my opinion this is bad practice 🙂.

Reply:
Fair enough. I'm not familiar with Swift so just putting ideas out but couldn't really comment on which one is best/worse from a code perspective here.

What're the other options do you think?

Enumerate the devices but for devices which are both input and output list them twice under different IDs? I.e. if device A has both input and output, it'd be listed in getAllDevices twice, under Ainput and Aoutput for example. Then either could be passed into toggleDeviceMute etc `toggleDeviceMute(AInput)`.

Other option would be to separate the methods out I suppose, i.e. toggleInputDeviceMute and toggleOutputDeviceMute? Could leave the existing implementation but add these to allow control of a specific device i/o.

Reply:
I think device with input and output channels will have same ID.
I would suggest in case of needing create separate methods to mute input or output channel and use `toggleInputMute` or `toggleOutputMute`.

Reply:
Makes sense to me. Would we need `getInputDeviceMute` and `getOutputDeviceMute` too to complete the api? Also, presumably `generateExport` will be needed for each?

Reply:
Done 😎